cmd/soroban-cli/src/events.rs

```diff
@@ -0,0 +1,437 @@
+            // though it's likely that this logic belongs more in
+            // `snapshot::read_events()`.
+            .rev()
+            // The ledger range is optional in sandbox mode.
+            .filter(|evt| {
+                if self.start_ledger == 0 && self.end_ledger == 0 {
+                    return true;
+                }
+
+                match evt.ledger.parse::<u32>() {
+                    Ok(seq) => seq >= self.start_ledger && seq <= self.end_ledger,
+                    Err(e) => {
+                        eprintln!("error parsing key 'ledger': {:?}", e);
+                        eprintln!(
+                            "your sandbox events file ('{}') may be corrupt",
+                            path.to_str().unwrap(),
+                        );
+                        eprintln!("ignoring this event: {:#?}", evt);
+
+                        false
+                    }
+                }
+            })
+            .filter(|evt| {
+                // Contract ID filter(s) are optional, so we should render all
+                // events if they're omitted.
+                self.contract_ids.is_empty()
+                    || self.contract_ids.iter().any(|id| *id == evt.contract_id)
+            })
+            .filter(|evt| {
+                // Like before, no topic filters means pass everything through.
+                self.topic_filters.is_empty() ||
+                // Reminder: All of the topic filters are part of a single
+                // filter object, and each one contains segments, so we need to
+                // apply all of them to the given event.
+                self.topic_filters
+                    .iter()
+                    // quadratic but both are <= 5 long
+                    .any(|f| does_topic_match(&evt.topic, &f.split(',').collect()))
+            })
+            .take(count)
+            .cloned()
+            .collect::<Vec<Event>>())
+    }
+}
+
+// Determines whether or not a particular filter matches a topic based on the
+// same semantics as the RPC server:
+//
+//  - for an exact segment match, the filter is a base64-encoded ScVal
+//  - for a wildcard, single-segment match, the string "*" matches exactly one
+//    segment
+//
+// The expectation is that a `topic_filter` is a comma-separated list of
+// segments that has previously been validated.
+//
+// [API
+// Reference](https://docs.google.com/document/d/1TZUDgo_3zPz7TiPMMHVW_mtogjLyPL0plvzGMsxSz6A/edit#bookmark=id.35t97rnag3tx)
+// [Code
+// Reference](https://github.com/stellar/soroban-tools/blob/bac1be79e8c2590c9c35ad8a0168aab0ae2b4171/cmd/soroban-rpc/internal/methods/get_events.go#L182-L203)
+pub fn does_topic_match(topics: &Vec<String>, segments: &Vec<&str>) -> bool {
+    let mut idx = 0;
+
+    for segment in segments {
+        if idx >= topics.len() {
+            // Nothing to match, need at least one segment.
+            return false;
+        }
+
+        if *segment != "*" {
```

Comment:
*bug*
Shouldn't this check if the segment *is* a wildcard?
```suggestion
        if *segment == "*" {
```

Could we add a couple unit-tests for this? Could just steal the test cases from the rpc server

Reply:
Done! Grabbed them all from RPC.